### PR TITLE
Added fire-and-forget Messaging protocol handler

### LINKED ISSUE #1202: Implement fire-and-forget messaging as a protocol handler

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

- Added protocol `/tari/messaging/0.1.0`
- This protocol handler handles inbound and outbound messaging.
- The node must already be connected to the peer before sending/receiving messages.
- Peers sending `/tari/messaging/0.1.0` messages between each other open
  two substreams (lazily, as needed), one for inbound and one for
  outbound.
- Outbound messages are sent as a request to MessagingProtocol. If a
  substream is not open for this peer, one is opened and 'cached' for
  future message sends.
- This protocol handler will be hooked up to the existing OMS and IMS.
  Essentially it is the glue between the existing fire-and-forget functionality
  and the async multiplexed connections.

## Motivation and Context

Closes #1202

## How Has This Been Tested?

New mocks for connection manager and peer connection. Wrote outbound, inbound and error case tests. 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/protocol/messaging/mod.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    PeerNotConnected,
+    #[error(no_from)]
+    UnexpectedConnectionManagerError(ConnectionManagerError),
+    /// Failed to send message
+    MessageSendFailed,
+    ProtocolError(ProtocolError),
+    PeerConnectionError(PeerConnectionError),
+}
+
+pub struct MessagingProtocol {
```
Comment (sdbondi) [inline comms/src/protocol/messaging/mod.rs]: 👁 s on this file

File: comms/src/protocol/messaging/mod.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+                },
+                Err(err) => {
+                    // Lazily remove Senders from the active queue if the MessagingProtocolHandler has shut down
+                    if err.is_disconnected() {
+                        self.active_queues.remove(&node_id);
+                        break;
+                    }
+
+                    attempts += 1;
+                    if attempts > 5 {
```
Comment (philipr-za) [inline comms/src/protocol/messaging/mod.rs]: This threshold should probably come from a const or a config

Comment (sdbondi) [reply comms/src/protocol/messaging/mod.rs]: True - I did kind of bang that retry functionality in there (after trying a [way to do it properly](https://gist.github.com/sdbondi/cbc43752faa75914b36ae9dd3fd13915)) - I think that this case should resolve pretty quickly or not happen at all so this is just a quick way to improve reliability on the chance that it happens. So I vote for a const in the file vs a configuration field - will update the code

File: comms/src/connection_manager/tests/manager.rs [hunk 3]
```diff
@@ -149,6 +149,13 @@ async fn simultaneous_dial_events() {
     result1.unwrap();
     result2.unwrap();
 
+    // Wait for listening and peer connected events
```
Comment (sdbondi) [inline comms/src/connection_manager/tests/manager.rs]: Fix for flakey test - wait until peer definitely has connected (i.e. after tie breaking) before shutting down

**Pull request merged**